### PR TITLE
Generate search data in ruby

### PR DESCRIPTION
**Pull request by bgracie**

### CODE REVIEW

File: lib/tasks/recache.rake [hunk 1]
```diff
@@ -117,16 +117,7 @@ namespace :pathways do
...
-        Specialization.all.each do |s|
-          puts "Search division #{d.id} specialization #{s.id}"
-          expire_fragment "livesearch_#{division_path(d)}_#{specialization_path(s)}_entries"
-
-          HttpGetter.exec("refresh_livesearch_division_entries/#{d.id}/#{s.id}.js")
-
-        end
-        expire_fragment "livesearch_#{division_path(d)}_content"
-        HttpGetter.exec("refresh_livesearch_division_content/#{d.id}.js")
+        division.search_data.regenerate_cache
```
Comment (KelseyDH) [inline lib/tasks/recache.rake]: shouldn't this be `d.search_data.regenerate_cache` ?

File: app/views/layouts/application.html.haml [hunk 2]
```diff
@@ -33,8 +41,8 @@
             search_all: '#search_all',
             search_all_url: '/livesearch_all_entries.js',
             global_data: pathways_global_search_data,
-            division_entry_data: pathways_division_search_data_entries,
-            division_content_data: pathways_division_search_data_content,
+            division_entry_data: pathways_divisional_entries_search_data,
+            division_content_data: pathways_divisional_content_search_data,
```
Comment (KelseyDH) [inline app/views/layouts/application.html.haml]: You renamed `pathways_division_search_data_entries` to `pathways_divisional_entries_search_data`

But `pathways_division_search_data_entries` is also called in *`views/search/_livesearch_division_entries.js.erb`* (line 49) and *`views/search/livesearch.js.haml`*  (line 15).  How are they affected by this change/refactor?

**Comment by KelseyDH** [comment]
Few thoughts:

1. Is this just a refactor or do your changes affect (recache) performance?

2. Have you thoroughly tested it on staging so that recaches work correctly?

**Comment by bgracie** [comment]
haven't tested it yet..

also, the dynamically generated js partials are going away, since the browser is improperly caching them.  (see https://github.com/pathwaysmedical/frasernw/issues/191)

File: lib/tasks/recache.rake (hunk 1 above)
Comment (bgracie) [inline lib/tasks/recache.rake]: yeah, good call!

**Comment by bgracie** [comment]
got this warning on the first recache of search, but it seemed to have still worked:
DalliError: Unable to unmarshal value: undefined class/module DivisionCity

- seems likely that my code triggered some cache read which returned a serialized DivisionCity, which the app hadn't loaded (for some reason)..
- more info here: http://alisdair.mcdiarmid.org/2013/02/02/fixing-rails-auto-loading-for-serialized-objects.html

**Comment by KelseyDH** [comment]
One advantage of using Get requests is that it meant we could distribute the load of recaching each fragment across our dynos on Heroku.  Do you know if this is still the case with your approach here, or is it all recaching one at a time?  I'm interested in how this change affects our recache performance.

**Comment by KelseyDH** [comment]
I guess another solution would be to delegate these recache tasks to Resque or Delayed Job, then schedule up a bunch of workers to process them during recaches.  Falling outside of Heroku's timeout limit would certainly be nice.

**Comment by bgracie** [comment]
In terms of time spent recaching, it doesn't matter that our web dynos (rather than the worker dyno runnning the recache task) are doing the heavy lifting in terms of cache generation.  The worker dyno is making those http requests synchronously, so it has to wait for responses....

**Comment by bgracie** [comment]
If you're concerned that the recache task is taking too long ( and I am too ), I would suggest trying to see which elements of it can be taken out and only done when the cache goes stale (i.e. when the relevant record is updated).  That approach also has the advantage of more consistently delivering up-to-date data to users.

**Pull request merged**